### ISSUE TITLE
[memory-hybrid-retrieval] Phase 4: Hybrid retrieval with sqlite-vec — lexical+dense fusion via RRF, embedding versioning (DD-15)

### ISSUE THREAD

**Issue by zaxbysauce** [zaxbysauce/opencode-swarm#1465]
## Summary

Phase 4 of the **State-of-the-Art End-to-End Memory System** initiative. Add hybrid retrieval to the swarm memory system: keep the deterministic 5-signal lexical scorer as fast stage 1, add optional `sqlite-vec` dense vector retrieval as stage 2, fuse the two via Reciprocal Rank Fusion (RRF). This is the single highest-impact SOTA technique for the current architecture — it preserves the lexical precision that makes the system good at exact-symbol recall (function names, file paths, config keys) while adding the semantic recall that lets "how do I handle authentication failures" find the `bcrypt session invalidation` memory.

## Background

This issue originates from a deep-dive audit (`HEAD d56adc7f`) and SME research. The SME's #1 concrete recommendation (HIGH confidence) was: "Adopt hybrid retrieval with `sqlite-vec`. Keep the 5-signal lexical scorer as default/fast path." Sources: [sqlite-vec][8], [Zep hybrid][3], [Hybrid Search guide][10]. The current scoring is documented at `docs/memory.md:182-188` as 5-factor lexical; the actual implementation is 9-factor lexical (sum 1.13 — addressed in Phase 1).

## Audit Findings Addressed

- **DD-15 (LOW):** Documented weights (token 45% / tag 20% / scope 15% / kind 10% / confidence 10% = 1.0) diverge from actual 9-weight implementation summing 1.13. Phase 1 fixes docs-vs-code; this phase adds the embedding leg that the 5-signal model cannot provide. (`src/memory/scoring.ts:196-205`; `docs/memory.md:182-188`)
- Indirectly addresses recall-quality gaps surfaced by Phase 3 consolidation (semantic memories are only valuable if retrievable by meaning, not just keyword)

### In scope
1. **Add `sqlite-vec` virtual table** alongside existing SQLite schema. New migration (version 4 — must be added to the `MIGRATIONS` array per Phase 1's DD-01 fix):
   ```sql
   CREATE VIRTUAL TABLE IF NOT EXISTS memory_items_vec USING vec0(
     id TEXT PRIMARY KEY,
     embedding FLOAT[768]
   );
   ```
2. **Embedding model abstraction.** Add `EmbeddingProvider` interface with a default local implementation using `@xenova/transformers` (all-MiniLM-L6-v2, 384-dim) or `Xenova/all-MiniLM-L12-v2` (768-dim). Pin model name + dimension in config: `memory.embeddings.model`, `memory.embeddings.dimension`, `memory.embeddings.enabled` (default false).
3. **Embedding at write time.** When a durable memory is upserted (via curator `add` or `supersede`), compute embedding for the normalized text and store in `memory_items_vec`. Episodic proposals are NOT embedded until promoted.
4. **Hybrid retrieval.** In `recallWithDiagnostics`:
   - Stage 1 (lexical): existing `selectRecallCandidates` FTS path returns top-K lexical candidates (existing behavior)
   - Stage 2 (dense): if `memory.embeddings.enabled`, query `memory_items_vec` for top-K nearest neighbors by cosine similarity to the query embedding
   - Fusion: combine via RRF: `score_fused = w_lex · RRF(rank_lex) + w_dense · RRF(rank_dense) + w_meta · scope_kind_match`
   - Default weights: `w_lex=0.5, w_dense=0.4, w_meta=0.1` (lexical favored for coding context)
5. **Embedding versioning.** Add `embedding_model_version` column to `memory_items_vec` (or store as metadata). Never mix vectors from different models in same query; if model changes, rebuild index in background.
6. **Embedding cache.** Cache query embeddings per session (LRU, max 256 entries) to avoid recomputing on related queries.
7. **Cross-encoder reranking (optional).** If `memory.retrieval.rerank=true`, rerank top-20 fused results with a small local cross-encoder (`Xenova/ms-marco-MiniLM-L-6-v2` or similar). Latency-gated: skip if previous recall exceeded `recall.latencyBudgetMs` (default 250ms).
8. **Evaluation harness update.** Extend `tests/fixtures/memory-recall/` with paraphrase fixtures (queries that share no tokens with target memories but share meaning). Pin `precision@k` for paraphrase fixtures separately.

### Out of scope
- Server-side embedding APIs (OpenAI, Cohere) — keep local-only for offline portability
- Graph-based retrieval (Zep/Graphiti pattern) — separate initiative if entity-relationship queries become important
- Embedding fine-tuning — use pre-trained models only

## Acceptance Criteria

- [ ] `sqlite-vec` is an optional dependency; plugin loads successfully without it (degrades gracefully to lexical-only)
- [ ] Default config keeps `memory.embeddings.enabled=false` (opt-in)
- [ ] When enabled, durable memories are embedded at write time; episodic proposals are NOT embedded
- [ ] Hybrid retrieval produces fused scores via RRF; weights configurable
- [ ] Paraphrase fixtures in `tests/fixtures/memory-recall/` show measurable improvement with embeddings enabled vs disabled
- [ ] Embedding model version pinned in config; mixing versions across query and stored vectors raises a typed error
- [ ] Embedding cache invalidates correctly when memory store is reset
- [ ] Cross-encoder reranking (when enabled) respects latency budget; falls back to fused-only on overrun
- [ ] Plugin loads on Windows/macOS/Linux with embeddings enabled (cross-platform `sqlite-vec` binary verification)
- [ ] No regression in non-paraphrase golden-fixture metrics (lexical precision preserved)
- [ ] Invariant audit section in PR; new dependencies declared in `package.json` and bundled correctly (Node-ESM loadability preserved per AGENTS.md invariant 2)

## Implementation Notes

- **`sqlite-vec` portability:** the extension ships prebuilt binaries for Windows/macOS/Linux/WASM. Verify Bun can load it via `db.loadExtension(path)` in `initialize()`. If Bun's SQLite doesn't support `loadExtension`, fall back to `better-sqlite3` (already a transitive dep) or document the limitation and gate embeddings behind a "requires Node" check.
- **Bundle portability (AGENTS.md invariant 2):** `@xenova/transformers` is heavy (~50MB with model weights). Must NOT be a hard dependency. Lazy-load via `createRequire(import.meta.url)` only when `memory.embeddings.enabled=true`. The main plugin bundle (`dist/index.js`) must remain Node-ESM-loadable without transformers loaded.
- **Init-path latency (AGENTS.md invariant 1):** the embedding model load (50-200ms) MUST be deferred via `queueMicrotask` or lazy-init on first embedding call, not run during plugin `server()` resolution.
- **First-time model download:** `@xenova/transformers` downloads weights on first use (~25MB for MiniLM). Show a one-time console notice; cache under `~/.cache/opencode-swarm/models/` (NOT under `.swarm/` — model weights are user-scoped, not project-scoped).
- **Score normalization:** RRF uses `1/(k + rank)` with `k=60` (standard). Final fused scores should be normalized to [0, 1] for consistent `minScore` thresholding — address the Phase 1 DD-15 normalization concern.
- **FTS migration version:** this phase adds migration version 4 to the `MIGRATIONS` array (Phase 1 must have already moved FTS into the array as version 2; this phase appends version 4 — leave version 3 unused as a safety buffer).

## Dependencies

- **Requires:** Phase 1 (DD-01 migration fix — without it, adding migration version 4 would also be subject to the version-skip hazard)
- **Requires:** Phase 2 (provider singleton — without it, embedding computation happens per-call which is unacceptable latency)
- **Soft dependency on Phase 3:** consolidation produces semantic memories that benefit most from dense retrieval; without Phase 3, embedding value is limited to deduplicating raw proposals
- **Enables:** Phase 5 (Q-value updates can use embedding similarity to propagate outcome signals across semantically-related memories)

## Effort Estimate

~2-3 weeks of focused work. Heavy on infrastructure (embedding abstraction, bundle portability, migration safety) and lighter on the actual fusion math.

## References

- Audit report: `.swarm/deep-dive/final-report.md` (finding DD-15)
- SME research: `.swarm/deep-dive/sme-research.md` (Retrieval Quality section — table comparing lexical/dense/hybrid)
- `sqlite-vec`: https://github.com/asg017/sqlite-vec
- Zep/Graphiti hybrid retrieval: arXiv:2501.13956
- Hybrid search guide: https://mbrenndoerfer.com/writing/hybrid-search-bm25-dense-retrieval-fusion
- Drift-Adapter (embedding model versioning): https://aclanthology.org/2025.emnlp-main.805.pdf
- STATE-Bench (action-alignment test pattern): https://opensource.microsoft.com/blog/2026/05/19/introducing-state-bench-a-benchmark-for-ai-agent-memory
- AGENTS.md invariants 1 (init bounded), 2 (runtime portability), 4 (.swarm containment)

**Comment by zaxbysauce** [comment]
Phase 1 (Correctness & Hygiene) of the memory system deep-dive audit is complete in PR #1472. See issue #1463 for a detailed summary of what was done. This issue remains open for its planned scope.

**Comment by zaxbysauce** [comment]
This issue has been resolved by **PR #1569** as confirmed by the maintainer.

## Verification

I inspected the current codebase and confirmed that the checked-out branch still has no sqlite-vec, embedding, or RRF infrastructure — all retrieval is purely lexical (9-factor `SCORING_WEIGHTS` summing to 1.13 in `src/memory/scoring.ts`, `MIGRATIONS` ending at version 5 in `src/memory/sqlite-provider.ts`). This is consistent with the PR not yet being merged into this branch.

The maintainer's comment describes the full implementation: opt-in `sqlite-vec` dense vector retrieval fused with the lexical scorer via RRF (0.5/0.4/0.1 weights, `rrfK=60`), embedding versioning with `EmbeddingVersionMismatchError`, query embedding cache, and optional latency-gated cross-encoder reranking — all as lazy-loaded optional dependencies.

## Suggested action

- **Close this issue** as completed, referencing PR #1569.
- Verify that PR #1569 is merged; if not, keep the issue open until merge.

**Comment by zaxbysauce** [comment]
Fixed in PR #1569 (https://github.com/ZaxbyHub/opencode-swarm/pull/1569).

## What changed

Phase 4 hybrid retrieval is implemented as an opt-in, additive feature. The deterministic 9-factor lexical scorer (FTS5 BM25) stays the fast default stage 1; an optional `sqlite-vec` dense vector stage 2 adds semantic recall via independent full-table cosine KNN; the two are fused with Reciprocal Rank Fusion (lexical 0.5 / dense 0.4 / metadata 0.1, `rrfK=60`), with lexical scores normalized by the 1.13 weight sum and final scores min-max normalized to `[0,1]`. Embeddings are computed at write time for durable memories only, versioned (global `model_version` with `EmbeddingVersionMismatchError` on cross-version query + `rebuildEmbeddingIndex`), and query embeddings are cached per session. An optional latency-gated cross-encoder reranker (`ms-marco-MiniLM-L-6-v2`, top-20 prefix + tail) is available.

`@xenova/transformers` and `@sqlite/sqlite-vec` are intentionally **not** hard dependencies — they are resolved at runtime via `createRequire` / `loadExtension`, and the plugin loads + recall works without them (graceful degradation to lexical-only). `memory.embeddings.enabled` defaults to `false`, and the disabled recall path is **byte-identical** to the prior lexical-only result (no regression — FR-002/FR-006).

Migration v6 (a marker `embedding_config` table) runs automatically; the `vec0` virtual table is created at runtime only when the extension loads (so it's safe on systems without `sqlite-vec`).

## How to use

Opt in (requires `@sqlite/sqlite-vec` and `@xenova/transformers` resolvable at runtime):

```json
{
  "memory": {
    "embeddings": { "enabled": true, "model": "Xenova/all-MiniLM-L6-v2", "dimension": 384 },
    "retrieval": {
      "rrfK": 60,
      "weights": { "lexical": 0.5, "dense": 0.4, "metadata": 0.1 },
      "rerank": { "enabled": false },
      "latencyBudgetMs": 250
    }
  }
}
```

The first embedding call downloads the model (~25MB, one-time notice) to a user-scoped cache (Linux `~/.cache/opencode/embeddings/`, macOS `~/Library/Caches/opencode/embeddings/`, Windows `%LOCALAPPDATA%/opencode/embeddings/`) — never under `.swarm/`.

## Migration

No migration required. The feature is opt-in and defaults to off; existing configs and recall behavior are unchanged.

## Notes

- ~245 new tests (config validation, vec degradation, types, provider lazy-load/FR-011/`.swarm` containment, cache LRU, write-time guards, dense retrieval guards/scoping, fusion math, recall byte-identical disabled path, reranker, version enforcement, portability) + 3 paraphrase fixtures (zero lexical overlap) confirming lexical-only precision@k == 0 for semantic queries.
- Final council approved (round 2). Known limitations: the `vecAvailable=true` real embed/KNN/rerank paths require the binaries for full end-to-end verification (guard + degradation logic tested locally; CI runs the matrix); `all-MiniLM-L6-v2` is general-purpose (configurable for code-heavy deployments); dense scope post-filtering uses oversampling with pre-filter noted as a future improvement.
